pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, shoot.Spec.SecretBindingName), shootSecretBinding); err != nil {
+		return nil, err
+	}
+	return kutil.GetSecretByRef(ctx, a.gardenClient.Client(), shootSecretBinding.SecretRef)
+}
+
+func (a *actuator) getShootKubeconfigSecret(ctx context.Context, shoot *gardencorev1beta1.Shoot) (*corev1.Secret, error) {
+	shootKubeconfigSecret := &corev1.Secret{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(shoot.Namespace, fmt.Sprintf("%s.kubeconfig", shoot.Name)), shootKubeconfigSecret); err != nil {
+		return nil, err
+	}
+	return shootKubeconfigSecret, nil
+}
+
+func (a *actuator) disableVPA(ctx context.Context, shoot *gardencorev1beta1.Shoot) (bool, error) {
+	if err := a.seedClient.Client().Get(ctx, kutil.Key(shoot.Status.TechnicalID, "vpa-admission-controller"), &appsv1.Deployment{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+const (
+	gardenletKubeconfigBootstrapSecretName = "gardenlet-kubeconfig-bootstrap"
+	gardenletKubeconfigSecretName          = "gardenlet-kubeconfig"
+)
+
+func (a *actuator) getGardenletChartValues(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) (map[string]interface{}, error) {
+	var err error
+
+	// Determine deployment values
+	var deploymentValues map[string]interface{}
+	if deploymentValues, err = utils.ToValues(managedSeed.Spec.Gardenlet.Deployment); err != nil {
+		return nil, err
+	}
+	if managedSeed.Spec.Gardenlet.MergeParentConfig {
+		parentDeployment, err := getParentDeployment(a.imageVector, shoot)
+		if err != nil {
+			return nil, err
+		}
+		parentDeploymentValues, err := utils.ToValues(parentDeployment)
+		if err != nil {
+			return nil, err
+		}
+		deploymentValues = utils.MergeMaps(parentDeploymentValues, deploymentValues)
+	}
+
+	// Determine config values
+	var configValues map[string]interface{}
+	if configValues, err = utils.ToValues(managedSeed.Spec.Gardenlet.Config); err != nil {
+		return nil, err
+	}
+	if managedSeed.Spec.Gardenlet.MergeParentConfig {
+		parentConfig, err := getParentConfig(a.config)
+		if err != nil {
+			return nil, err
+		}
+		parentConfigValues, err := utils.ToValues(parentConfig)
+		if err != nil {
+			return nil, err
+		}
+		configValues = utils.MergeMaps(parentConfigValues, configValues)
+	}
+
+	// Marshal config values back to an object
+	var configObj *configv1alpha1.GardenletConfiguration
+	if err := utils.FromValues(configValues, &configObj); err != nil {
+		return nil, err
+	}
+
+	// If a bootstrap mechanism is specified, compute bootstrap kubeconfig and set gardenClientConnection values accordingly
+	// Otherwise, if kubeconfig path is specified in gardenClientConnection, read it and store its contents
+	var gccValues map[string]interface{}
+	if gccValues, err = utils.GetMapFromValues(configValues, "gardenClientConnection"); err != nil {
```

Comment:
can it be nil?

Reply:
In the code below this line, `gccValues` are manipulated in a way that can't be done in `configObj`, since the Helm chart values contains fields such as `bootstrapKubeconfig.kubeconfig` that are not part of `GardenletConfiguration` at all.

`gccValues` could happen to be `nil` in theory which is not an issue for reading. The following line makes sure it's not nil before we try to set anything:

```
gccValues = utils.InitValues(gccValues)
```

Reply:
ok thanks